velox/functions/sparksql/Map.cpp

```diff
@@ -32,50 +32,42 @@
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-template <TypeKind kind>
 void setKeysResultTyped(
     vector_size_t mapSize,
     std::vector<VectorPtr>& args,
     const VectorPtr& keysResult,
     exec::EvalCtx& context,
     const SelectivityVector& rows) {
-  using T = typename KindToFlatVector<kind>::WrapperType;
-  auto flatVector = keysResult->asFlatVector<T>();
-  flatVector->resize(rows.end() * mapSize);
-
   exec::LocalDecodedVector decoded(context);
+
   for (vector_size_t i = 0; i < mapSize; i++) {
     decoded.get()->decode(*args[i * 2], rows);
-    // For efficiency traverse one arg at the time
+    SelectivityVector targetRows(keysResult->size(), false);
+    std::vector<vector_size_t> toSourceRow(keysResult->size());
     rows.applyToSelected([&](vector_size_t row) {
       VELOX_CHECK(!decoded->isNullAt(row), "Cannot use null as map key!");
-      flatVector->set(row * mapSize + i, decoded->valueAt<T>(row));
+      targetRows.setValid(row * mapSize + i, true);
+      toSourceRow[row * mapSize + i] = row;
     });
+    targetRows.updateBounds();
+    keysResult->copy(args[i * 2].get(), targetRows, toSourceRow.data());
   }
 }
 
-template <TypeKind kind>
 void setValuesResultTyped(
     vector_size_t mapSize,
     std::vector<VectorPtr>& args,
     const VectorPtr& valuesResult,
-    exec::EvalCtx& context,
     const SelectivityVector& rows) {
-  using T = typename KindToFlatVector<kind>::WrapperType;
-  auto flatVector = valuesResult->asFlatVector<T>();
-  flatVector->resize(rows.end() * mapSize);
-
-  exec::LocalDecodedVector decoded(context);
   for (vector_size_t i = 0; i < mapSize; i++) {
-    decoded.get()->decode(*args[i * 2 + 1], rows);
-
+    SelectivityVector targetRows(valuesResult->size(), false);
```

Comment:
This logic seems the same as in setKeyXxx function. Would it make sense to combine these two functions to avoid repeated computations and copy-paste?

Reply:
Make sense, updated.